src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1626,19 +1640,36 @@ public override BoundNode VisitCall(BoundCall node)
             {
                 (ParameterSymbol parameter, _) = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded);
                 AttributeAnnotations annotations = parameter?.FlowAnalysisAnnotations ?? AttributeAnnotations.None;
+                // We'll ignore annotations that are misused
+
+                // NotNullWhenTrue (without NotNullWhenFalse) must be applied on a bool-returning member
+                if ((annotations & AttributeAnnotations.NotNullWhenTrue) != 0 &&
+                    (annotations & AttributeAnnotations.NotNullWhenFalse) == 0 &&
+                    parameter.ContainingSymbol.GetTypeOrReturnType().SpecialType != SpecialType.System_Boolean)
+                {
+                    annotations &= ~AttributeAnnotations.NotNullWhenTrue;
+                }
 
-                // We'll ignore NotNullWhenFalse that is misused in metadata
+                // NotNullWhenTrue must be applied to a reference or unconstrained generic type
+                if ((annotations & AttributeAnnotations.NotNullWhenTrue) != 0 &&
+                    (parameter.Type?.IsValueType != false || parameter.IsParams))
+                {
+                    annotations &= ~AttributeAnnotations.NotNullWhenTrue;
+                }
```

Comment:
Consider extracting a helper method for both misused `NotNullWhenTrue` and `NotNullWhenFalse`.